nrf_rpc/nrf_rpc.c

```diff
@@ -0,0 +1,786 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#define NRF_RPC_LOG_MODULE NRF_RPC
+#include <nrf_rpc_log.h>
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <string.h>
+
+#include "nrf_rpc.h"
+#include "nrf_rpc_tr.h"
+
+#define ERROR_FLAG_REMOTE 1
+#define ERROR_FLAG_SEND 2
+
+#define CMD_FLAG_WITH_RSP 0x10000
+
+enum {
+	/* Packet types with header variant 1:
+	 *      byte 0: bit 7: packet type, bits 6-0: group id,
+	 *      byte 1: command/event id
+	 */
+	PACKET_TYPE_CMD = 0x00,
+	PACKET_TYPE_EVT = 0x80,
+
+	/* Packet types with header variant 2:
+	 *      byte 0: packet type
+	 *      byte 1: 0xFF,
+	 */
+	PACKET_TYPE_RSP = 0x01,
+	PACKET_TYPE_ACK = 0x02,
+	PACKET_TYPE_ERR = 0x03,
+};
+
+
+/* Array with all defiend groups */
+NRF_RPC_AUTO_ARR(nrf_rpc_groups_array, "grp");
+
+
+/* ======================== Common utilities ======================== */
+
+
+/** Put header variant 1 into packet based on specified parameters */
+static inline void header_variant_1(uint8_t *packet, uint8_t type,
+				    uint8_t group_id, uint8_t id)
+{
+	packet[0] = type | group_id;
+	packet[1] = id;
+}
+
+
+/** Put header variant 2 into packet based on specified parameters */
+static inline void header_variant_2(uint8_t *packet, uint8_t type)
+{
+	packet[0] = type;
+	packet[1] = 0xFF;
+}
+
+
+/** Function simplifying sending a short packets */
+static int send_simple(struct nrf_rpc_tr_local_ep *tr_local_ep,
+		       struct nrf_rpc_tr_remote_ep *tr_remote_ep, uint8_t type,
+		       const uint8_t *packet, size_t len)
+{
+	uint8_t *tx_buf;
+
+	nrf_rpc_tr_alloc_tx_buf(tr_remote_ep, &tx_buf,
+				_NRF_RPC_HEADER_SIZE + len);
+
+	if (nrf_rpc_tr_alloc_failed(tx_buf)) {
+		return NRF_RPC_ERR_NO_MEM;
+	}
+
+	header_variant_2(tx_buf, type);
+
+	if (packet != NULL && len > 0) {
+		memcpy(&tx_buf[_NRF_RPC_HEADER_SIZE], packet, len);
+	}
+
+	return nrf_rpc_tr_send(tr_local_ep, tr_remote_ep, tx_buf,
+			       _NRF_RPC_HEADER_SIZE + len);
+}
+
+
+/** Report an error that cannot be reported as a function return value */
+static void report_error(struct nrf_rpc_tr_local_ep *tr_local_ep,
+			 struct nrf_rpc_tr_remote_ep *tr_remote_ep, int code,
+			 uint32_t flags)
```

Comment:
> Is there anything the user can do to recover from transport level errors in a deployment?

No, but the user may want to know about it to take some actions, e.g. reset the device, put it in some different log, show it on UI.

> TCP doesn't have special behavior for treating HTTP-error codes, as that is communicated on a higher layer.

But in application that have both TCP and HTTP there can be one common error handler for both of them. Maybe it is a little confusing that I put error handler in a specific layer. Actually it is common for all layers. "ep" parameters are related to the lower level (transport), but going back to web application you can pass socket on which HTTP error occurred to the error handler (socket describes TCP connection).

I can consider separating error handling if you think that it will be more clear approach? I think that creating two error handling systems adds to much unnecessary complexity for the user (and also in the code).

Reply:
Based on your suggestions, I did more error separation (e.g. cmd/evt/rsp handlers are not allowed to pass errors to the nRF PRC). In case of this function I have to keep it in the API, but I added a following doxygen comment:

```
 * Main intention for exposing this function is to allow serialization layer
 * (e.g. TinyCBOR) report error to the same error handler as entire nRF RPC.
```